File: musesp_ui/components/scroll_list.py
"""ScrollList 可滚动列表组件。"""

from collections.abc import Callable

import pygame

from musesp_ui.components.core import Component


class ScrollList(Component):
    """可滚动的列表组件。

    每项为调用方构造的 Component（需设置 _item_id 属性），
    通过 set_items() 设置。
    """

    def __init__(self, x: int = 0, y: int = 0, width: int = 0, height: int = 0,
                 item_height: int = 32):
        """:param item_height: 每项固定高度"""
        super().__init__(x=x, y=y, width=width, height=height)
        self.item_height = item_height
        self._scroll = 0
        self._max_scroll = 0
        self._on_select: Callable[[str], None] | None = None
        self._selected_id: str | None = None

    def set_items(self, items: list[Component]) -> None:
        """设置列表项。每项需设置 ``_item_id`` 作为标识。"""
        self._sub_components.clear()
        for i, comp in enumerate(items):
            comp.height = self.item_height
            self._sub_components.append(comp)
        self._max_scroll = max(
            0, len(items) * self.item_height - self.height)
        self._scroll = min(self._scroll, self._max_scroll)
        self._update_positions()

    def layout(self) -> None:
        super().layout()
        for child in self._sub_components:
            self._propagate_width(child, self.width)

    @staticmethod
    def _propagate_width(comp: Component, width: int) -> None:
        comp.width = width
        for sub in comp.sub_components:
            ScrollList._propagate_width(sub, width)

    def bind_on_select(self, handler: Callable[[str], None]) -> None:
        """绑定选中回调。``handler(item_id)``。重复点击已选中项不触发。"""
        self._on_select = handler

    # ---- 滚动位移直接写入子组件真实 y，默认绘制/调试/事件逻辑无需改动 ----

    def _update_positions(self) -> None:
        for i, child in enumerate(self._sub_components):
            child.y = i * self.item_height - self._scroll

    # ---- 绘制：仅需裁剪 ----

    def _draw_self(self, surface: pygame.Surface, draw_x: int, draw_y: int) -> None:
        pygame.draw.rect(surface, (30, 30, 30),
                         pygame.Rect(draw_x, draw_y, self.width, self.height))
        # 滚动条（仅在可滚动时显示）
        if self._max_scroll > 0:
            total_h = len(self._sub_components) * self.item_height
            bar_w = 4
            bar_h = max(20, self.height * self.height // total_h)
            track_h = self.height - bar_h
            bar_y = draw_y + (self._scroll * track_h // self._max_scroll
                              if self._max_scroll > 0 else 0)
            bar_x = draw_x + self.width - bar_w - 2
            pygame.draw.rect(surface, (100, 100, 100),
                             (bar_x, bar_y, bar_w, bar_h),
                             border_radius=bar_w // 2)

    def _draw_internal(self, surface: pygame.Surface,
                       offset_x: int, offset_y: int) -> None:
        draw_x = self.x + offset_x
        draw_y = self.y + offset_y
        self._draw_self(surface, draw_x, draw_y)
        old_clip = surface.get_clip()
        surface.set_clip(
            pygame.Rect(draw_x, draw_y, self.width, self.height))
        for child in self._sub_components:
            if -child.height < child.y < self.height:
                child._draw_internal(surface, draw_x, draw_y)
        surface.set_clip(old_clip)

    # ---- 事件：仅需可见性过滤 ----

    def dispatch_event(self, event: pygame.event.Event) -> bool:
        local = self._shift_event(event, -self.x, -self.y)
        if not self._handle_event(local):
            return False
        for child in self._sub_components:
            if -child.height < child.y < self.height:
                if not child.dispatch_event(local):
                    return False
        return True

    def _handle_event(self, event: pygame.event.Event) -> bool:
        if not super()._handle_event(event):
            return False
        if event.type == pygame.MOUSEWHEEL:
            self._scroll -= event.y * 24
            self._scroll = max(0, min(self._scroll, self._max_scroll))
            self._update_positions()
        elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            if self._in_rect(event.pos) and self._on_select:
                for child in self._sub_components:
                    if child.y <= event.pos[1] < child.y + child.height:
                        item_id = getattr(child, "_item_id", "")
                        if item_id != self._selected_id:
                            self._selected_id = item_id
                            self._on_select(item_id)
                        return False
        return True

File: musesp_ui/src/components/scroll_list.rs
use std::future::Future;
use std::pin::Pin;

use winit::event::{ElementState, MouseButton, MouseScrollDelta, WindowEvent};

use async_trait::async_trait;

use crate::components::core::{ComponentBase, ComponentTrait, Direction};
use crate::renderer::UIRenderer;

pub type SelectHandler = Box<dyn FnMut(&str) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send>;

pub struct ScrollList {
    pub base: ComponentBase,
    pub item_height: i32,
    pub scroll: i32,
    pub max_scroll: i32,
    pub on_select: Option<SelectHandler>,
    pub selected_id: Option<String>,
}

impl ScrollList {
    pub fn new(x: i32, y: i32, width: i32, height: i32, item_height: i32) -> Box<Self> {
        Box::new(ScrollList {
            base: ComponentBase::new(x, y, width, height),
            item_height,
            scroll: 0,
            max_scroll: 0,
            on_select: None,
            selected_id: None,
        })
    }

    pub fn set_items(&mut self, items: Vec<Box<dyn ComponentTrait>>) {
        self.base.children.clear();
        let count = items.len() as i32;
        for mut item in items {
            item.base_mut().height = self.item_height;
            self.base.children.push(item);
        }
        self.max_scroll = (count * self.item_height - self.base.height).max(0);
        self.scroll = self.scroll.min(self.max_scroll);
        self.update_positions();
        self.propagate_width();
    }

    pub fn bind_on_select(&mut self, handler: SelectHandler) {
        self.on_select = Some(handler);
    }

    fn update_positions(&mut self) {
        for (i, child) in self.base.children.iter_mut().enumerate() {
            child.base_mut().y = i as i32 * self.item_height - self.scroll;
        }
    }

    fn propagate_width(&mut self) {
        let w = self.base.width;
        Self::propagate_width_recursive(&mut self.base.children, w);
    }

    fn propagate_width_recursive(children: &mut [Box<dyn ComponentTrait>], width: i32) {
        for child in children {
            child.base_mut().width = width;
            Self::propagate_width_recursive(child.base_mut().children.as_mut_slice(), width);
        }
    }
}

#[async_trait]
impl ComponentTrait for ScrollList {
    fn base(&self) -> &ComponentBase {
        &self.base
    }
    fn base_mut(&mut self) -> &mut ComponentBase {
        &mut self.base
    }
    fn draw_self(&self, renderer: &mut UIRenderer, dx: i32, dy: i32) {
        renderer.draw_rect(dx, dy, self.base.width, self.base.height, (30, 30, 30, 255));
        let total = self.base.children.len() as i32 * self.item_height;
        let max_scroll = (total - self.base.height).max(0);
        if max_scroll > 0 {
            let bar_w = 4i32;
            let bar_h = (self.base.height * self.base.height / total).max(20);
            let track_h = self.base.height - bar_h;
            let bar_y = if max_scroll > 0 {
                dy + self.scroll * track_h / max_scroll
            } else {
                dy
            };
            let bar_x = dx + self.base.width - bar_w - 2;
            renderer.draw_rect(bar_x, bar_y, bar_w, bar_h, (100, 100, 100, 255));
        }
    }

    fn draw(&self, renderer: &mut UIRenderer, offset_x: i32, offset_y: i32) {
        let dx = self.base.x + offset_x;
        let dy = self.base.y + offset_y;
        self.draw_self(renderer, dx, dy);
        // 裁剪子组件绘制到列表区域内，对齐 Python surface.set_clip
        renderer.push_clip(dx, dy, self.base.width, self.base.height);
        for child in &self.base.children {
            let cy = child.base().y;
            if -child.base().height < cy && cy < self.base.height {
                child.draw(renderer, dx, dy);
            }
        }
        renderer.pop_clip();
    }

    fn layout(&mut self, h_override: Option<Direction>) {
        self.base.do_layout(h_override);
        self.propagate_width();
    }

    async fn dispatch_event(&mut self, event: &WindowEvent) -> bool {
        match event {
            WindowEvent::MouseWheel { delta, .. } => {
                let (lx, ly) = self.base.local_pos(self.base.cursor_x, self.base.cursor_y);
                if self.base.in_rect(lx, ly) {
                    let scroll_amount = match delta {
                        MouseScrollDelta::LineDelta(_, y) => (*y * 24.0) as i32,
                        MouseScrollDelta::PixelDelta(pos) => pos.y as i32,
                    };
                    let new_scroll = (self.scroll - scroll_amount).clamp(0, self.max_scroll);
                    if new_scroll != self.scroll {
                        self.scroll = new_scroll;
                        self.update_positions();
                    }
                    return true;
                }
            }
            WindowEvent::CursorMoved { device_id, position } => {
                let (lx, ly) = self.base.local_pos(position.x, position.y);
                if !self.base.handle_mouse_move(lx, ly, event).await {
                    return false;
                }
                let local_event = WindowEvent::CursorMoved {
                    device_id: *device_id,
                    position: winit::dpi::PhysicalPosition::new(lx as f64, ly as f64),
                };
                return self.dispatch_to_visible_children(&local_event).await;
            }
            WindowEvent::MouseInput { state, button, .. } => {
                let (lx, ly) = self.base.local_pos(self.base.cursor_x, self.base.cursor_y);

                if *state == ElementState::Pressed && *button == MouseButton::Left {
                    if self.base.in_rect(lx, ly) {
                        let mut triggered: Option<String> = None;
                        for child in &self.base.children {
                            let cy = child.base().y;
                            if -child.base().height < cy && cy < self.base.height {
                                if child.base().in_rect(lx, ly - cy) {
                                    if let Some(ref item_id) = child.base().item_id {
                                        if self.selected_id.as_deref() != Some(item_id) {
                                            triggered = Some(item_id.clone());
                                        }
                                    }
                                    break;
                                }
                            }
                        }
                        if let Some(ref item_id) = triggered {
                            self.selected_id = Some(item_id.clone());
                            if let Some(ref mut handler) = self.on_select {
                                handler(item_id).await;
                            }
                            return false;
                        }
                    }
                }

                if !self.base.handle_mouse_input(*state, *button, lx, ly, event).await {
                    return false;
                }
                return self.dispatch_to_visible_children(event).await;
            }
            WindowEvent::KeyboardInput {
                event: key_event, ..
            } => {
                if !self.base.handle_keyboard(key_event, event).await {
                    return false;
                }
                return self.dispatch_to_visible_children(event).await;
            }
            _ => return self.dispatch_to_visible_children(event).await,
        }
        true
    }

    fn set_scroll_items(&mut self, items: Vec<Box<dyn ComponentTrait>>) {
        self.set_items(items);
    }
}

impl ScrollList {
    async fn dispatch_to_visible_children(&mut self, event: &WindowEvent) -> bool {
        let n = self.base.children.len();
        let local_cx = self.base.cursor_x - self.base.x as f64;
        let local_cy = self.base.cursor_y - self.base.y as f64;
        for i in 0..n {
            let cy = self.base.children[i].base().y;
            if -self.base.children[i].base().height < cy && cy < self.base.height {
                self.base.children[i].base_mut().cursor_x = local_cx;
                self.base.children[i].base_mut().cursor_y = local_cy;
                if !self.base.children[i].dispatch_event(event).await {
                    return false;
                }
            }
        }
        true
    }
}
